authorization.go

```diff
@@ -0,0 +1,204 @@
+	// in it's `httprouter.Params`. In which case the value of this parameter is passed
+	// as `uriSite`.
+	// The values of `clientOU` and `clientCN` are the same as those of `IsAuthorized`.
+	IsAuthorizedForSite(clientOU []string, clientCN, uriSite string) error
+}
+
+// AllowOUs is a convenience function which produces an AuthorizationCheck from a list
+// of allowed OUs. See AllowSpecificOUandCNs for behavioral details.
+func AllowOUs(allowedOUs []string) AuthorizationCheck {
+	return AllowSpecificOUandCNs{OUs: allowedOUs, CNs: []string{}}
+}
+
+// AllowOUsWithSiteRestrictions is a convenience function which produces
+// two AuthorizationChecks which restrict all endpoints to a list of valid OUs,
+// and additional implement site-resource checks for a subset of those OUs.
+// AllowedOUs specifies which OUs a request must be part of to be serviced.
+// RestrictedSiteOUs specifies which OUs to apply site-resource checks to.
+// See AllowSpecificOUandCNs and RestrictSiteAccess for behavioral details on both
+// checks.
+// For example:
+//
+//   opts := certauth.Opts{
+//       AuthorizationChecks: AllowOUsWithSiteRestrictions(
+//           []string{"admin", "site", "employee"},
+//           []string{"site"},
+//       ),
+//   }
+//
+// This example would require *all* requests to present an OU of `admin`, `site`, or
+// `employee` and additionally, requests with an OU of `site` would be prevented from
+// accessing site-specific resources that were not part of their own site.
+func AllowOUsWithSiteRestrictions(
+	AllowedOUs, RestrictedSiteOUs []string,
+) []AuthorizationCheck {
+	return []AuthorizationCheck{
+		AllowOUs(AllowedOUs),
+		RestrictSiteAccess{AppliedOUs: RestrictedSiteOUs},
+	}
+}
+
+// AllowSpecificOUandCNs is an AuthorizationCheck which allows access to
+// resources for the specific Organizational Units and common names.
+// If any of the client's OUs match (i.e. `==`) any of the server's allowed OUs, *and*
+// the client's CN matches (i.e. `==`) one of the server's allowed CNs, the request
+// is allowed.
+// If `OUs` is empty, the client's OU is ignored, and only the CN is used to determine
+// authorization.
+// If `CNs` is empty, the client's CN is ignored, and only the OU is used to determine
+// authorization.
+// If both `OUs` and `CNs` are empty, no authorization check is performed.
+// Site resources are not considered specially. IsAuthorizedForSite has exactly the same
+// behavior as IsAuthorized (i.e. the site is ignored).
+type AllowSpecificOUandCNs struct {
+	OUs []string
+	CNs []string
+}
+
+func (allow AllowSpecificOUandCNs) IsAuthorized(
+	clientOU []string, clientCN string,
+) error {
+	if len(allow.OUs) > 0 {
+		if err := allowedOU(allow.OUs, clientOU); err != nil {
+			return err
+		}
+	}
+	if len(allow.CNs) > 0 {
+		if err := allowedCN(allow.CNs, clientCN); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (allow AllowSpecificOUandCNs) IsAuthorizedForSite(
```

Comment:
prefix `Is` kind of implies return of `bool` ... How about `AuthorizeForSite` and `allow.Authorize`?

Reply:
> prefix Is kind of implies return of bool

That's fair. I think I'll rename these methods `CheckAuthorization` and `CheckAuthorizationForSite`. I like that better than `Authorize...` cause the latter feels like it has side-effects when it really doesn't.